### DISCUSSION TITLE
Does +r1 +r2 +r3 in the version mean build?

### DISCUSSION THREAD

**Discussion by cdwmhcc** [deskflow/deskflow#7658] [Q&A] [answered]
Dear development team:

What does the +r in the +r1 +r2 +r3 versions stand for? Is it build?

The current version numbers include: 
- 1.16.1-beta+r1
- 1.16.1+r1
- 1.16.1
- 1.14.6.19-stable
- 1.14.6.16-rc
- 1.13.1.40-rc1
- etc.

**Comment by sithlord48** [comment, marked as answer]
soon nothing i plan to redo how our versions are done soon.

for now every pr that lands makes a new r# tag why idk ....it was that way before.

**Reply by cdwmhcc** [reply]
Thanks for your reply and explanation, I understand now.

**Reply by nbolton** [reply]
> for now every pr that lands makes a new r# tag why idk ....it was that way before.

`r` = revision, as in "n revisions since the patch was bumped"

In the old method, using `+` meta was to make the version number strict semver... but we don't really need that.

**Reply by SelfRef** [reply]
I'd like to have a request for the future to not change the tag (especially the latest one) that already exists. That breaks AUR scripts I'm maintaning. Thanks in advance.

**Reply by cdwmhcc** [reply]
> Using he `+` meta was to make the version number strict semver

In fact, it does not follow the SemVer MAJOR.MINOR.PATCH-prerelease+build rule: https://github.com/deskflow/deskflow/releases/tag/1.14.6.17-rc

**Reply by cdwmhcc** [reply]
> I'd like to have a request for the future to not change the tag (especially the latest one) that already exists.

I am also maintaining automation scripts and have found that I need to customize rules specifically for the version numbers of deskflow. The current naming convention increases the burden of maintenance and understanding. Adopting the SemVer standard could reduce this burden and enhance transparency and consistency in versioning, for example:

- `1.16.1-beta.1`
- `1.14.6-rc.17`
- `1.16.5`

Such a naming convention is clear, intuitive, and easy to understand, which helps with the compatibility and efficiency of automation tools. I believe standardizing version numbers offers great convenience for developers who are currently maintaining or will maintain automation scripts in the future, and it also makes it much easier for regular users to understand the differences between versions.

I hope we can consider this improvement to create a better experience for all users and developers.

**Reply by nbolton** [reply]
> I'd like to have a request for the future to not change the tag

@SelfRef Sorry for the disruption. Teething problems while we move away from the Synergy way of doing things to a new era of Deskflow.

> The current naming convention

@cdwmhcc By 'current' do you mean @sithlord48's new convention (`major.minor.patch.tweak`) or the old convention with `+`?

Edit: Ah, I guess you mean [`v1.17.0`](https://github.com/deskflow/deskflow/releases/tag/v1.17.0) which should probably be `v1.17.0.0`, right?

**Reply by SelfRef** [reply]
@nbolton No problem. I'm fully aware there will be changes and the project is yet to be estabilished. I only assumed that such changes can be done safer while creating a new releases, because that won't break existing links.

For now it's not a big deal, because there's no other packages than my Arch ones as I see.

**Reply by nbolton** [reply]
FWIW, old tags are available in the downstream Synergy fork: https://github.com/symless/synergy-deskflow

**Reply by cdwmhcc** [reply]
> By 'current' do you mean @sithlord48's new convention (`major.minor.patch.tweak`) or the old convention with `+`?

When I posted this reply, sithlord48 had not posted the following reply yet:

> We use a new naming convention
> Major . minor. patch .tweak
> where M.m.p match the last tag (vM.m.p) and Tweak is the number of commits since the last tag this will be the version scheme after https://github.com/deskflow/deskflow/pull/7672 lands.


So here I am referring to the old convention with +

**Comment by sithlord48** [comment]
We use a new naming convention 
Major . minor. patch .tweak 

where M.m.p match the last tag (vM.m.p) and Tweak is the number of commits since the last tag this will be the version scheme after #7672  lands.

@SelfRef  sorry the the new tags will be in the form of vM.m.p this will not change anymore.

**Reply by cdwmhcc** [reply]
Absolutely, regardless of the naming convention used, the most important aspect is consistency.

As you mentioned with the format `major.minor.patch.tweak`:
- 1.16.1.34
- 1.16.1.33

It's crucial not to mix different schemes, which could lead to confusion, such as:
- 1.16.2
- 1.16.1.36

**Reply by SelfRef** [reply]
Thanks. The new naming convention is definitely better.

**Reply by nbolton** [reply]
> the most important aspect is consistency

I agree with this. Please pardon the disruption while we transition to a new convention. Instead of `1.16.2`, I believe it would be `1.16.2.0` (for the first "tweak") -- right @sithlord48?

**Reply by cdwmhcc** [reply]
> I agree with this. Please pardon the disruption while we transition to a new convention. Instead of `1.16.2`, I believe it would be `1.16.2.0` (for the first "tweak") -- right @sithlord48?

Yes, there won't be 1.6.0, there will be 1.6.0.0

The convention `major.minor.patch.tweak` established by @sithlord48 seems reasonable if it fits well with your team's approach, yet I wonder if it might be the best option available. How does this versioning scheme handle pre-release labels like beta, alpha, or RC? Moreover, I'm curious about the choice not to implement [SemVer](https://semver.org/). Could you please provide some insights into these decisions?

**Reply by nbolton** [reply]
> I'm curious about the choice not to implement [SemVer](https://semver.org/)

I believe @sithlord48 will be the best to answer this. I initially felt that using [SemVer](https://semver.org/) (e.g. `1.16.2-1` or `1.16.2-beta.1` instead of `1.16.2.1`) was the way to go but we concluded together that we didn't need to use SemVer (we couldn't think of a good reason since we're not a library), so we settled on the 4-digit non-SemVer format. I say that the 4-digit style is non-SemVer based on [semver.org](https://semver.org/):

> "A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version."

I belive the term "tweak" as we use in our new style is interchangeable with what SemVer calls "pre-release version".

It's possibly not too late to change that decision though and go with SemVer... but the window is closing.

> How does this versioning scheme handle pre-release labels like beta, alpha, or RC?

We may decide not to use pre-release labels such as those since we have a leading-edge philosophy (as such every release is likely to be akin to beta). @sithlord48 and I haven't explicitly discussed pre-release labels yet, so it would be good to hear his views on this.

**Reply by cdwmhcc** [reply]
I think there might be some misunderstanding regarding the SemVer. It is important to clarify that SemVer is not exclusively designed for libraries; its core purpose is to provide a standardized method of version numbering. As I mentioned before, I maintain an automation tool that requires comparison of version numbers. If the versioning isn't standardized, we're forced to rely on complex and error-prone regex matching.

Additionally, regarding the version numbering format, although there may be historical reasons for using the `major.minor.patch.tweak` format, this method is not very flexible and tends to be cumbersome. Especially when dealing with pre-release versions like beta, rc, or alpha, the four-digit format does not scale well.

SemVer offers a more flexible naming convention: `MAJOR.MINOR.PATCH-prerelease+build(metadata)`. For example:
`1.16.2-alpha.1` < `1.16.2-beta.1` < `1.16.2-beta.2+3` < `1.16.2-rc.1` = `1.16.2-rc.1+88` < `1.6.2`, and they clearly indicate the order of releases. the +build is purely an internal build metadata, which is not significant to the end-user but provides valuable build information to the development team without causing version confusion.

Therefore, I suggest reconsidering the adoption of SemVer. It could not only help us manage versions more effectively but also prevent some potential issues in the future.

**Reply by nbolton** [reply]
@cdwmhcc You make a compelling argument. Hmm, ok now I'm leaning toward SemVer in favor of the non-SemVer 4-digit style. We're on Matrix BTW, if you prefer real-time discussion: https://matrix.to/#/#deskflow:matrix.org

@sithlord48 What are your thoughts? Would SemVer be practical for us to use? I think now is the time to change our mind on this if we want to. It'll be tricky to change this later.

**Reply by sithlord48** [reply]
- Beta and Rc are not valid in A Continuous workflow
   - Every release is either `Stable` or not
   - `-git` packages would track continuous
   - if you run the app and the last digit is not 0 its not "stable". and the tweak makes it easy to see how off stable you are
- Stable tag must be i the form or `vM.m.p` (i.e 1.17.1) 
   - Every version is unique once the code hits master
   - Simple numeric comparisons
   - The app will always report a proper version
 
- Windows requires this format for the version scheme in rc files. so now linux windows, mac all show the same 4 point version

**Comment by nbolton** [comment]
> I am also maintaining automation scripts 

Are you able to share the project/system name that you're maintaining the scripts for?

**Reply by sithlord48** [reply]
he is talking about these https://aur.archlinux.org/packages?O=0&K=deskflow (they all apper to be fixed other then -bin broken for other reasons)

**Reply by SelfRef** [reply]
That was not my message with automation scripts. It was from @cdwmhcc.

I'm only maintaining the Arch AUR packages through my [WIP AUR semi-automatic portal](https://github.com/SelfRef/auromatic).

**Comment by nbolton** [comment]
It appears that this discussion has concluded.

To wrap up, it looks like we're sticking with the 4-digit style and not SemVer.
